library/common/jni/jni_interface.cc

```diff
@@ -143,22 +143,23 @@ static JNIEnv* get_env() {
   return env;
 }
 
-static void jvm_on_headers(envoy_headers headers, bool end_stream, void* context) {
+static void* jvm_on_headers(envoy_headers headers, bool end_stream, void* context) {
   JNIEnv* env = get_env();
   jobject j_context = static_cast<jobject>(context);
+  pass_headers(env, headers, j_context);
 
   jclass jcls_JvmCallbackContext = env->GetObjectClass(j_context);
   jmethodID jmid_onHeaders = env->GetMethodID(jcls_JvmCallbackContext, "onHeaders", "(JZ)V");
   // Note: be careful of JVM types. Before we casted to jlong we were getting integer problems.
   // TODO: make this cast safer.
-  env->CallVoidMethod(j_context, jmid_onHeaders, (jlong)headers.length,
-                      end_stream ? JNI_TRUE : JNI_FALSE);
+  jobject result = env->CallObjectMethod(j_context, jmid_onHeaders, (jlong)headers.length,
+                                         end_stream ? JNI_TRUE : JNI_FALSE);
 
   env->DeleteLocalRef(jcls_JvmCallbackContext);
-  pass_headers(env, headers, j_context);
+  return result;
```

Comment:
As far as I can tell the returned contexts are not getting used anywhere yet right?

Reply:
Correct, they're used in the subsequent PR I have for filters.